### PR TITLE
Validate setters

### PR DESCRIPTION
**Pull request by jdhiguita**
## Summary
- Add preconditions to user and book models
## Trello Card
https://trello.com/c/H0y0udQY

### CODE REVIEW

File: src/main/java/wolox/training/models/Book.java [hunk 1]
```diff
@@ -57,6 +63,7 @@ public String getGenre() {
     }
 
     public void setGenre(String genre) {
+        Preconditions.checkArgument(genre != null && !genre.isEmpty(), "Genre must have a valid value");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: You can replace this check with

```
import org.apache.commons.lang.StringUtils;
 StringUtils.isEmpty ()
```
It is a method that performs both checks.

```suggestion
        Preconditions.checkArgument(!StringUtils.isEmpty(genre), "Genre must have a valid value");
```
You could take all the sets that have a string as a parameter, delete the checkNotNull and place this check.

File: src/main/java/wolox/training/models/Book.java [hunk 2]
```diff
@@ -105,6 +117,7 @@ public String getYear() {
     }
 
     public void setYear(String year) {
+        Preconditions.checkNotNull(year, "Year cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: In addition to the change suggested above, we could add a check that verifies that the content of this string is numeric

`.checkArgument (StringUtils.isNumeric (year), "error message");`

File: src/main/java/wolox/training/models/Book.java [hunk 3]
```diff
@@ -121,6 +135,7 @@ public String getIsbn() {
     }
 
     public void setIsbn(String isbn) {
+        Preconditions.checkNotNull(isbn, "Isbn cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: In addition to the change suggested above, we could add a check that verifies that the content of this string is numeric

`.checkArgument (StringUtils.isNumeric (isbn), "error message");`

File: src/main/java/wolox/training/models/Book.java [hunk 4]
```diff
@@ -113,6 +126,7 @@ public int getPages() {
     }
 
     public void setPages(int pages) {
+        Preconditions.checkNotNull(pages, "Pages cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: Here we could add a check that controls if the number entered in the pages is greater than 0

File: src/main/java/wolox/training/models/User.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    public long getId() {
+        return id;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void setUsername(String username) {
+        Preconditions.checkNotNull(username, "Username cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: You can use the following method:

```
import org.apache.commons.lang.StringUtils;
 StringUtils.isEmpty ()
```
It is a method that performs both a check for null and for empty.

```suggestion
        Preconditions.checkArgument(!StringUtils.isEmpty(username), "Username cannot be null or empty");
```

File: src/main/java/wolox/training/models/User.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        this.username = username;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        Preconditions.checkNotNull(name, "Name cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: You can use the following method:

```
import org.apache.commons.lang.StringUtils;
 StringUtils.isEmpty ()
```
It is a method that performs both a check for null and for empty.

```suggestion
        Preconditions.checkArgument(!StringUtils.isEmpty(name), "Name cannot be null or empty");
```

File: src/main/java/wolox/training/models/User.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        this.name = name;
+    }
+
+    public LocalDate getBirthdate() {
+        return birthdate;
+    }
+
+    public void setBirthdate(LocalDate birthdate) {
+        Preconditions.checkNotNull(birthdate, "Birthdate cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: Add a check that the date of birth is before today's date.

```suggestion
        Preconditions.checkArgument(! birthdate.isAfter(LocalDate.now()), "message");
```

File: src/main/java/wolox/training/models/Book.java [hunk 8]
```diff
@@ -81,6 +90,7 @@ public String getTitle() {
     }
 
     public void setTitle(String title) {
+        Preconditions.checkNotNull(title, "Title cannot be null");
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: I recommend using a class to store error messages as constants. The best thing would be to create a package called constants and in there host your class of constants for example ErrorConstants.

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]
Great progress, leave you just a few comments 😄

File: src/main/java/wolox/training/models/Book.java (hunk 1 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/Book.java]: Update preconditions to setters with string argument

File: src/main/java/wolox/training/models/Book.java (hunk 2 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/Book.java]: Add new validate to set

File: src/main/java/wolox/training/models/Book.java (hunk 3 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/Book.java]: Add new validate to set

File: src/main/java/wolox/training/models/Book.java (hunk 4 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/Book.java]: Add new validate greater than 0 in set

File: src/main/java/wolox/training/models/User.java (hunk 5 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/User.java]: Update precondition type in set

File: src/main/java/wolox/training/models/User.java (hunk 6 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/User.java]: Update precondition type in set

File: src/main/java/wolox/training/models/User.java (hunk 7 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/User.java]: Add new validate to set

File: src/main/java/wolox/training/models/Book.java (hunk 8 above)
Comment (jdhiguita) [reply src/main/java/wolox/training/models/Book.java]: Create class with constants to error messages

**Review by Gianni-Nicolas — APPROVED** [review]

**Review by cramallo — APPROVED** [review]

**Pull request merged**